### ISSUE TITLE
FileHandle: Shared Storage User Permissions - Android 11 (API 30)

### ISSUE THREAD

**Issue by Tekkerue** [libgdx/libgdx#6755]
**ISSUE**

gdx.files.FileHandle does not appear to be able to handle user permissions when accessing Shared Storage on Android 11 (API 30). The Absolute FileType can be used to create and manipulate files in Shared Storage (I am using the public Documents directory) while the app is installed. However, when the app is uninstalled and reinstalled Android treats the new installation as a different app and will not allow access to files created by the previous installation unless permission is obtained first.

My app allows users to create and save files, but I do not want these files to be automatically deleted when the app is uninstalled, so I cannot use the Android/data directory. That would be like Photoshop deleting all of your Photoshop documents and rendered images if it is uninstalled, which is terrible! There must be a way to save user created files and then regain access to them again when the app is reinstalled.

**POTENTIAL SOLUTION**

FileHandle should provide a setting to be able to work with files on Shared Storage and also prompt for user permission to access them when necessary.

A response to one of my questions on Stack Overflow suggested that LibGDX should use URI’s instead of file paths. They said that once permission is granted to access a file, the permission is for the URI, not the file path. Therefore, even if you obtain a file path from the URI, you lose the permission since it is stored with the URI. The MediaStore is another method I’ve seen for accessing files in Shared Storage, would it be possible to use this on the Android backend? Another potential option is “File Path API” which I believe uses MediaStore under the hood, but I’m not sure on this as I don’t really understand it.

I have made multiple attempts to get something work with the FileHandle, but I’ve hit a dead end every time because my knowledge of Android programming is very limited (I have focused on learning LibGDX which abstracts the native Android code). It might be possible to get this working with FileHandle, but it is above my programming abilities, especially for native Android. I hope the experts who work on LibGDX will consider adding in this functionality to the Android backend. Thanks!

**Comment by yajirobe69** [comment]
fwiw "but I do not want these files to be automatically deleted when the app is uninstalled" you supposedly can use this flag in the manifest to prevent that: `android:hasFragileUserData="true"`

**Comment by Tekkerue** [comment]
Thank you for the reply. The hasFragileUserData flag was mentioned on the LibGDX discord, but they said "that does not work well at the time being but you should set it anyway".

I have plans for additional features like allowing users to load in audio files, export audio files and maybe MIDI (if I can figure this out). Having a folder off of the root like I originally planned is a lot faster to access, but since I can't do that now the public Documents folder will be fine. From my understanding this is doable, but it requires using Android's new shared storage system on the Android backend. I have seen a few potential options (Storage Access Framework, MediaStore, and using URI's instead of file paths), but I don't know which of these would be best to use.

I've tried asking on Stack Overflow and the LibGDX discord and I think I'm going to have to make a duplicate FileHandle and try to figure out the Android backend myself. I have absolutely no idea what I'm doing in Android and this is probably going to take forever to figure out. I'm getting very close to beta release with my app, for now I'll just put a warning to let people know they'll need to manually make a copy of these files before uninstalling the app if they want to keep them (and hope people remember they need to do this). Then I'll just slowly chip away at it and see if I can figure it out. I'm not sure what else I can do.

Thanks again.

**Comment by marekhalmo** [comment]
Hello,.. had you any luck with this? 
I'm struggling with the same issue.

My app needs to unpack zip from download folder and put it to shared storage - the data needs to stay after uninstall..

I also tried to duplicate FileHandle class, same for applications createFiles() method so i get my own FileHandle implementation every time i requrest FileHandle..
So far no luck :(

This Android 11 restriction is crazy.

**Comment by Tekkerue** [comment]
@marekhalmo no, I have not. I tried looking into it, but ultimately gave up as I'm not an Android programmer. I was hoping people more knowledgeable of native Android would fix this for LibGDX, but so far that hasn't happened yet. My best solution is just to put a notice on the save screen to let users know where preset files are saved and that they should back the folder up before uninstalling the app. I'm not happy with that, but I have no idea what else to do.

**Comment by marekhalmo** [comment]
@Tekkerue after a loooooong looong development i got something to work. It is fairly complicated.
I had to override DefaultAndroidFiles.getFileHandle(String path, FileType type) method (and others) to create a document file handle..
in short: After you get your Uri from the app permission request - you need to convert it to an AndroidDocument based file handle. Every operation with files from that point has to go from this android document (which basically gives you permissions).. so no constructing of file paths - you have to go via FileHandle parent/child/list methods..

I have found no clean solution for this yet and it is still fairly buggy..

Good luck!

**Comment by Tekkerue** [comment]
@marekhalmo Wow, really? Would you mind sharing the code for the files you modified and a working example showing how to use it? If you submit it to the LibGDX team maybe they would implement it officially into LibGDX. Sounds like you created a good starting point, even if it's a little buggy. I made a few different attempts myself but couldn't get anything working and finally gave up.

**Comment by MobiDevelop** [comment]
This is a fairly nuanced problem that might be difficult to solve without accepting that it will necessarily act differently to other file handles. That is, some concepts of FileHandle don't map well to the SAF way and some of the conveniences that we have when using File backed handles don't apply.

Reading and writing single documents, for which you have permissions, is relatively straightforward. Dealing with a document tree is where it gets more complicated as you have to start with a tree url and traversing the tree is really only possible one step at a time.

I've put together a basic implementation of a DocumentFileHandle that can handle most of the important operations, which I am happy to share (maybe as an extension or something), after I clean it up and test it a bit more.

**Comment by marekhalmo** [comment]
@MobiDevelop sharing that would be great. I would gladly contribute to your solution. My solution is not clean enough to share + I'm still struggling with performance (If you list files of a Tree document and then try to access properties like name e.t.c. there will be a new query executed for every instance of file in that list which makes the access fairly slow - i have a solution for that). 

So thanx for your answer. Please share your code so we can colaborate. Maybe it will go to libgdx ..

**Comment by marekhalmo** [comment]
Maybe a fork of libgdx would be great as well :)

**Comment by MobiDevelop** [comment]
Ha, yeah, efficiency isn't what DocumentFile is about - it's definitely targeting convenience. It's all a balance, I suppose.

DocumentFile goes to the content provider for all attributes, which ensures you'll always get up to date info, but can be slow. If it's too slow, you can probably cache attributes with the small risk of being outdated should something else change the underlying document. It sort of depends on the use case, e.g. a file picker has a much different scope than just reading/writing known files.

**Comment by marekhalmo** [comment]
@Tekkerue i don't think this would be possible.
Shared storage gives you an opportunity to create a different home folder for your app with access to read/write files.
That should be possible. 
I'm not sure if it will give you access to documents folder. 
Maybe you are trying to add a functionality to open specific kind of document? That would be something else..

**Comment by Tekkerue** [comment]
@marekhalmo In my original comment, I mentioned I didn't want to us the Android/data directory (shared storage) because the files are deleted when the app is uninstalled. The preset files I'm saving are saved as .json text files, so nothing too complicated. This used to work fine prior to Android 11, but now nothing works.

I was able to save/read files from the public documents folder initially, but if I uninstall and reinstall the app then I no longer have access to those files because the new installation was viewed as a "different" app. I wish there was a way to simply prompt the user to take ownership of that folder again, then I could continue using it as before. I don't understand why Android had to screw this up so badly, it shouldn't be difficult to ask the user before accessing a folder and then retain the permission so you only have to ask once.

I've got more ideas for feature and apps allowing users to add their own audio files (I'm developing apps for musicians) and would prefer to share these files between apps instead of having multiple copies in each app's data directory since music files are larger files than text files and storage space can be very low on mobile devices.

At the moment the preset manager is the feature that is currently broken, but Android's file system is also messing up features and apps that I'm considering making in the future. I have not been able to find a way to work around any of this yet, and it's very frustrating.
